python/ray/train/function_trainer.py

```diff
@@ -0,0 +1,121 @@
+import inspect
+from typing import Any, Optional, Callable, Type
+
+from ray.air.checkpoint import Checkpoint
+from ray.air.config import RunConfig, ScalingConfig
+from ray.train.trainer import BaseTrainer
+from ray.tune.trainable import Trainable, wrap_function
+from ray.tune.utils import detect_checkpoint_function, detect_config_single
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class FunctionTrainer(BaseTrainer):
+    """Trainer for generic (non-distributed) training functions.
+
+    This trainer can be used to wrap a generic function trainable. In contrast
+    to e.g. the DataParallelTrainer, this trainer will only start up one worker
+    (the trainer). Starting up potential distributed workers is then up
+    to the custom training function.
+
+    The trainer accepts a ``train_fn``, which can be any Ray Tune-compatible
+    function trainable.
+
+    If a scaling config is given, resources will be requested for all workers.
+    Starting potential distributed workers within the current placement group
+    is up to the user.
+
+    If a checkpoint to resume from is given, it will be passed as the
+    ``checkpoint_dir`` argument of the training function, if supported.
+    Alternatively, ``ray.air.session.get_checkpoint()`` can be used.
+
+    This trainer does not handle datasets and preprocessors. If this is needed,
```

Comment:
I think we should keep datasets and preprocessors right? The dataset should be accessible via session.get_dataset_shard().

We don't have to split the dataset, but we should still fit the BaseTrainer API, no?

Reply:
cc @xwjiang2010

Reply:
+1, think that makes sense

Reply:
I've originally left this out to not tinker with the user-provided checkpoints. But I guess we can add this back and document that if you use preprocessors, it will add them to your checkpoint.

Reply:
So I'm in the process of adding it, but `train.get_dataset_shard()` requires a train session to be set up. I'd specifically like to avoid that as we want to provide a layer to run (tune) function trainables without a train layer. Starting a train session would lead us to using a BackendExecutor managing another layer of workers (or rather a single worker), etc. It generally introduces a lot of complexity just for the sake of being able to specify resources.

I think there are some possible solutions here:

1. Don't use datasets and preprocessors. Users can pass them via `tune.with_parameters` if needed or construct them in their training function. Arguably you don't need our session-guided sharding mechanisms if you're using a single worker per default.
2. Add something like `session.get_dataset()` to return the full dataset. This would have to be added to both the tune and train session, but could be guarded on the AIR level. Not ideal to clutter the tune session scope imo
3. Pass them in the `config` dict to the `train_fn` if provided
4. Add them as optional parameters to the `train_fn`, only pass them if provided (similar to the old `checkpoint_dir`)

Thinking more about this, I think the config dict approach is least invasive. Also when users are using `Tuner(train_fn, param_space={"preprocessor": tune.choice([pp1, pp2])}` it could be expected that these are passed as config arguments.

Reply:
RE: `tune and train session`, aren't they the same thing for the users now?

Reply:
RE: "tune and train session, aren't they the same thing for the users now?" - they are the same in the sense of reporting and loading checkpoint
However, they are just different in some aspects - e.g. get_xxx may make sense for train session and not make sense for tune session.
My personal opinion is be explicit and use train.get_dataset_shard, train.get/set_accelerator etc and not complicate session with that. At one point, I even thought to scope session solely around reporting and loading checkpoint (the commonality). I know we have session.get_world_size() etc here right now. TBH I am still doubtful if that's the best idea for the exact reason why Richard has this confusion. What Amog and I said last time is this is not high priority and we can wait to see some user feedback to decide. 

Now regarding this thread, should we separate API requirement of this FunctionTrainer aside from its implementation details of it?
Let's first agree on what functionality FunctionTrainer should provide? And this should have nothing to do with what functionality we want Session to provide?

Reply:
Got it; sorry that makes sense. All I was suggesting was that we don't introduce new ways to access datasets and preprocessors when we don't need to. 

From the user perspective, the following is desirable.

1. I import FunctionTrainer from Train.
2. When using a FunctionTrainer, all Train functionality should be available to me. Including datasets and preprocessing and tuning.
3. I don't need to learn a new way to access the dataset, or learn a new way to pass my trainer into Tuner.

Now that being said, it may be the case that this is not easy to do right now, and this is relatively uncommon functionality. How about just tagging FunctionTrainer as alpha and adding a note that this API may change in the future?

Reply:
Agreed- my preference would be to access dataset through `session.get_dataset_shard()` for consistency with the other Trainers.

Reply:
Agreed. I would rather us not have this class, rather than have a half-baked FunctionTrainer. If we add this, then it should be a "real" Trainer. Otherwise just tell users to use Tune by itself with the old API.

Reply:
Is `session.get_dataset_shard()` the only missing requirement for the FunctionTrainer?

Reply:
ah got it. 
Yeah agree that API wise we should aim for having full Trainer functionality.
Implementation wise, we can have `session.get_dataset()` to be something meaningful for both train and tune sessions.
How does that sound?

Reply:
> Starting a train session would lead us to using a BackendExecutor managing another layer of workers (or rather a single worker), etc. It generally introduces a lot of complexity just for the sake of being able to specify resources.

Actually isn't this the right approach? It should be the same as num_workers=0 for SKlearntrainer, etc. It's actually simpler to use the same executor in all cases, rather than try to special case in this scenario.